src/backends/gitlab/API.js

```diff
@@ -0,0 +1,196 @@
+  // of the cursors, since GitLab's pagination sorts the opposite way
+  // we want to sort by default (it sorts by filename _descending_,
+  // while the CMS defaults to sorting by filename _ascending_, at
+  // least in the current GitHub backend). This should eventually be
+  // refactored.
+  listFiles = async path => {
+    const firstPageCursor = await this.fetchCursor({
+      url: `${ this.repoURL }/repository/tree`,
+      params: { path, ref: this.branch },
+    });
+    const lastPageLink = firstPageCursor.data.getIn(["links", "last"]);
+    const { entries, cursor } = await this.fetchCursorAndEntries(lastPageLink);
+    return { files: entries.reverse(), cursor: this.reverseCursor(cursor) };
+  };
+
+  traverseCursor = async (cursor, action) => {
+    const link = cursor.data.getIn(["links", action]);
+    const { entries, cursor: newCursor } = await this.fetchCursorAndEntries(link);
+    return { entries: entries.reverse(), cursor: this.reverseCursor(newCursor) };
+  };
+
+  listAllFiles = async path => {
+    const entries = [];
+    let { cursor, entries: initialEntries } = await this.fetchCursorAndEntries({
+      url: `${ this.repoURL }/repository/tree`,
+      // Get the maximum number of entries per page
+      params: { path, ref: this.branch, per_page: 100 },
+    });
+    entries.push(...initialEntries);
+    while (cursor && cursor.actions.has("next")) {
+      const link = cursor.data.getIn(["links", "next"]);
+      const { cursor: newCursor, entries: newEntries } = await this.fetchCursorAndEntries(link);
+      entries.push(...newEntries);
+      cursor = newCursor;
+    }
+    return entries;
+  };
+
+  toBase64 = str => Promise.resolve(Base64.encode(str));
+  fromBase64 = str => Base64.decode(str);
+  uploadAndCommit = async (item, { commitMessage, updateFile = false, branch = this.branch }) => {
+    const content = await (item instanceof AssetProxy ? item.toBase64() : this.toBase64(item.raw));
+    const file_path = item.path.replace(/^\//, "");
+    const action = (updateFile ? "update" : "create");
+    const encoding = "base64";
+    const body = JSON.stringify({
+      branch,
+      commit_message: commitMessage,
+      actions: [{ action, file_path, content, encoding }],
+    });
+
+    await this.request({
+      url: `${ this.repoURL }/repository/commits`,
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body,
+    });
+
+    return { ...item, uploaded: true };
+  };
+
+  persistFiles = (files, { commitMessage, newEntry }) => console.log(files, commitMessage, newEntry) ||
+    Promise.all(files.map(file => this.uploadAndCommit(file, { commitMessage, updateFile: newEntry === false })));
+
+  deleteFile = (path, commit_message, options = {}) => {
+    const branch = options.branch || this.branch;
+    return flow([
+      unsentRequest.fromURL,
+      unsentRequest.withMethod("DELETE"),
+      unsentRequest.withParams({ commit_message, branch }),
+      this.requestJSON,
```

Comment:
This endpoint returns a 204 and no content if the deletion was successful. We are trying to parse JSON here, which crashes on an empty string ("Unexpected end of JSON input.").